src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+
+  let chunk_off_to_poff c off = Int63.Syntax.(off - c.start_off)
+
+  type create_error = Io.create_error
+
+  type open_error =
+    [ Io.open_error
+    | `Closed
+    | `Invalid_argument
+    | `Inconsistent_store
+    | `Read_out_of_bounds ]
+
+  (** A simple container for chunks. *)
+  module Inventory : sig
+    type t
+
+    val v : int -> (int -> chunk) -> t
+    val iter_skip_appendable : (chunk -> unit) -> t -> unit
+    val appendable : t -> chunk
+    val find : off:int63 -> t -> chunk
+
+    val open_ :
+      start_idx:int ->
+      chunk_num:int ->
+      open_chunk:
+        (chunk_idx:int -> is_appendable:bool -> (Ao.t, open_error) result) ->
+      (t, [> open_error ]) result
+  end = struct
+    type t = chunk Array.t
+
+    exception OpenInventoryError of open_error
+
+    let v = Array.init
+
+    let iter_skip_appendable f t =
+      Array.sub t 0 (Array.length t - 1) |> Array.iter f
+
+    let appendable t = Array.get t (Array.length t - 1)
+
+    let find ~off t =
+      let find c =
+        let end_poff = Ao.end_poff c.ao in
+        let is_after_start = c.start_off <= off in
+        let is_before_end = chunk_off_to_poff c off < end_poff in
+        is_after_start && is_before_end
+      in
+      match Array.find_opt find t with
+      | None -> raise (Errors.Pack_error `Read_out_of_bounds)
+      | Some c -> c
+
+    let open_ ~start_idx ~chunk_num ~open_chunk =
+      let off = ref Int63.zero in
+      let create_chunk i =
+        let start_off = !off in
+        let chunk_idx = start_idx + i in
+        let is_appendable = i = chunk_num - 1 in
+        let open_result = open_chunk ~chunk_idx ~is_appendable in
+        match open_result with
+        | Error err -> raise (OpenInventoryError err)
+        | Ok ao ->
+            let end_poff = Ao.end_poff ao in
+            (off := Int63.Syntax.(!off + end_poff));
+            { idx = chunk_idx; start_off; ao }
+      in
+      try Ok (v chunk_num create_chunk)
+      with OpenInventoryError err ->
+        Error (err : open_error :> [> open_error ])
+  end
+
+  type t = { inventory : Inventory.t }
+
+  (* A lightweight wrapper around creating/opening append only files as chunks *)
+  module Ao_chunk = struct
+    let path = Layout.V4.suffix_chunk
+
+    let create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure =
+      let path = path ~root ~chunk_idx in
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
+      let path = path ~root ~chunk_idx in
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      let dead_header_size = if is_appendable then dead_header_size else 0 in
```

Comment:
I'm not sure why we need this check. Isn't it the case that `dead_header_size` is 0 for all files created by V3 and above?

Reply:
I ignored `dead_header_size` at first too but the tests failed (ty tests!). The reason it is needed for the "appendable" (ie last) chunk is that this will be the chunk for migrated v1/v2 stores that have never been (or cannot be -- ie are non-minimal) GCed. The same thing is happening for rw chunks but it is less obvious because there is no check since they are (by definition) "appendable".

Reply:
Aha! You're right. Previously, I was always using `dead_header_size=0`, which broke tests, but I confused myself when I applied the fix. Just using the passed in `dead_header_size` does work. Thanks for the investigation!